### PR TITLE
feat: ability to set dedupe globally for `io`

### PR DESCRIPTION
**Pull request by amitksingh1490**
**Summary:**  
_Briefly describe the changes made in this PR._

**Issue Reference(s):**  
Fixes #... _(Replace "..." with the issue number)_

**Build & Testing:**

- [ ] I ran `cargo test` successfully.
- [ ] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [ ] I have added relevant unit & integration tests.
- [ ] I have updated the [documentation] accordingly.
- [ ] I have performed a self-review of my code.
- [ ] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

### CODE REVIEW

File: tests/core/snapshots/async-cache-enabled.md_merged.snap [hunk 1]
```diff
@@ -4,7 +4,7 @@ expression: formatter
 ---
 schema
   @server(port: 8000, queryValidation: false)
-  @upstream(baseURL: "http://jsonplaceholder.typicode.com", dedupe: true) {
+  @upstream(baseURL: "http://jsonplaceholder.typicode.com", dedupe: "min") {
```
Comment (tusharmath) [inline tests/core/snapshots/async-cache-enabled.md_merged.snap:7]: Users should get a compile error when the value is not an enum variant.

File: src/core/async_cache.rs [hunk 2]
```diff
@@ -42,37 +79,92 @@ impl<Key: Eq + Hash + Send + Clone, Value: Debug + Clone + Send, Error: Debug +
...
+                    CacheValue::Pending(tx) => Some(tx.subscribe()),
+                    CacheValue::Ready(value) => unimplemented!(),
+                }
+            } else {
+                None
             }
+        };
+
+        if let Some(mut subscriber) = subscriber {
+            subscriber.recv().await.unwrap()
```
Comment (meskill) [inline src/core/async_cache.rs:127]: I changed this part of the code so lock is hold now until we get the result of `tx.subscribe()`.

I believe in previous version of the code when we've been doing cloning of the `tx` there was small period of time when lock is released, but subscriber is not yet created. And if meantime the value is resolved and send to the subscribers then this current subscriber that will be created will never get any value at all leading to deadlock

File: generated/.tailcallrc.graphql [hunk 3]
```diff
@@ -350,10 +350,20 @@ directive @upstream(
...
+  When set to `true`, it will ensure no HTTP, GRPC, or any other IO call is made more 
+  than once if similar request is inflight within the context of the Application.
+  """
+  dedupeInFlight: Boolean
+  """
   When set to `true`, it will ensure no HTTP, GRPC, or any other IO call is made more 
   than once within the context of a single GraphQL request.
   """
-  dedupe: Boolean
+  dedupeInRequest: Boolean
```
Comment (tusharmath) [inline generated/.tailcallrc.graphql:366]: Need to update the documentation

File: ci-benchmark/nginx-benchmark.graphql [hunk 4]
```diff
@@ -5,6 +5,7 @@ schema
     poolMaxIdlePerHost: 200
     tcpKeepAlive: 60
     proxy: {url: "http://127.0.0.1:3000"}
+    dedupeInFlight: true
```
Comment (tusharmath) [inline ci-benchmark/nginx-benchmark.graphql:8]: Let's keep this disabled for this benchmark

File: src/core/http/request_context.rs [hunk 5]
```diff
@@ -34,7 +34,8 @@ pub struct RequestContext {
     pub min_max_age: Arc<Mutex<Option<i32>>>,
     pub cache_public: Arc<Mutex<Option<bool>>>,
     pub runtime: TargetRuntime,
-    pub cache: AsyncCache<u64, ConstValue, EvaluationError>,
+    pub cache: AsyncCache<u64, ConstValue, EvaluationError, Cache>,
+    pub async_loader: Arc<AsyncCache<u64, ConstValue, EvaluationError, NoCache>>,
```
Comment (tusharmath) [inline src/core/http/request_context.rs:38]:
```suggestion
    pub cache_global: AsyncCache<u64, ConstValue, EvaluationError, NoCache>,
```

File: generated/.tailcallrc.graphql [hunk 6]
```diff
@@ -360,6 +360,16 @@ directive @upstream(
   """
   dedupe: Boolean
```
Comment (tusharmath) [inline generated/.tailcallrc.graphql:361]: This setting should be on the server.

File: generated/.tailcallrc.graphql [hunk 7]
```diff
@@ -360,6 +360,16 @@ directive @upstream(
...
   """
+  When set to `true`, it will ensure no graphQL execution is made more than once if 
+  similar query is being executed within the context of the Application.
```
Comment (tusharmath) [inline generated/.tailcallrc.graphql:364]:
```suggestion
  than once if similar request is inflight across the server's lifetime.
```

File: src/core/ir/io.rs [hunk 8]
```diff
@@ -64,22 +64,50 @@ impl Eval for IO {
...
+                    (false, true) => {
+                        ctx.request_ctx
+                            .dedupe_handler
+                            .dedupe(&key, || Box::pin(self.eval_inner(ctx)))
+                            .await
+                    }
+                    (false, false) => unreachable!(), // This case is handled at the beginning
                 }
             })
         } else {
```
Comment (tusharmath) [inline src/core/ir/io.rs:109]: Handle default case over here also.

File: src/core/ir/io.rs [hunk 9]
```diff
@@ -64,22 +64,50 @@ impl Eval for IO {
         &'a self,
         ctx: super::EvaluationContext<'a, Ctx>,
     ) -> Pin<Box<dyn Future<Output = Result<ConstValue, EvaluationError>> + 'a + Send>> {
-        if ctx.request_ctx.upstream.dedupe {
+        // Note: Handled the case separately to avoid cache key generation when it's not
+        // required
```
Comment (tusharmath) [inline src/core/ir/io.rs:68]:
```suggestion
        // Note: Handled the case separately for performance reasons. It avoids cache key generation when it's not
        // required
```

**Review by tusharmath — APPROVED** [review]

**Pull request merged**